### PR TITLE
Return constraint metadata with violations

### PR DESCRIPTION
**Pull request by morgante**
This PR adds the constraint metadata into the output returned with violations.

### CODE REVIEW

File: api/validator.proto [hunk 1]
```diff
@@ -67,11 +74,22 @@ message AuditResponse {
 message ResetRequest {}
 message ResetResponse {}
 
+message GetConstraintRequest {
+  // The name of the constraint to fetch
```
Comment (t12g) [inline api/validator.proto]: what about group/version/kind? We can probably assume the group and version, but kind needs to be part of the request, right?

File: pkg/gcv/cf/cf.go [hunk 2]
```diff
@@ -231,3 +231,18 @@ func (cf *ConstraintFramework) Audit(ctx context.Context) (*validator.AuditRespo
 
 	return response, nil
 }
+
+// GetConstraint returns a constraint by name
+func (cf *ConstraintFramework) GetConstraint(ctx context.Context, name string) (*validator.Constraint, error) {
```
Comment (t12g) [inline pkg/gcv/cf/cf.go]: can we add unit test for this? (also include a case for same name appearing under 2 different kinds)

File: api/validator.proto (hunk 1 above)
Comment (morgante) [reply api/validator.proto]: We don't provide the constraint kind as part of the violation data, so we'd need to add that.

**Comment by morgante** [comment]
> I'm probably missing the context... what's the motivation behind this PR?

The context is mainly to provide additional user feedback in violation reports by including annotations and other information from the original constraint. Specifically, for the scorecard CLI I was [prototyping](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/113).

One alternative would be to provide the full constraint config with every violation, but that would end up being a lot of data. Which approach do you prefer?

**Comment by t12g** [comment]
> > I'm probably missing the context... what's the motivation behind this PR?
> 
> The context is mainly to provide additional user feedback in violation reports by including annotations and other information from the original constraint. Specifically, for the scorecard CLI I was [prototyping](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/113).
> 
> One alternative would be to provide the full constraint config with every violation, but that would end up being a lot of data. Which approach do you prefer?

Providing the full constraint config with every violation can potentially generate a lot of redundant data. I prefer the current approach. Would we want to examine both constraint and template?

**Comment by morgante** [comment]
> Providing the full constraint config with every violation can potentially generate a lot of redundant data. I prefer the current approach. Would we want to examine both constraint and template?

From the admin perspective, I think the template is actually an implementation detail so I'm don't have a need for it.

Instead of including kind, could we enforce a requirement that constraint names be unique across kinds?

**Comment by t12g** [comment]
> From the admin perspective, I think the template is actually an implementation detail so I'm don't have a need for it.
> 
> Instead of including kind, could we enforce a requirement that constraint names be unique across kinds?

We can if constraint framework is planning to impose this requirement across the board. Otherwise it's a bit odd to have different restrictions between gatekeeper and config validator. 

Hmm... Maybe including constraint content in the violation data is an easier route at the cost of violation data bloat. 

@briantkennedy @maxsmythe your thoughts?

**Comment by briantkennedy** [comment]
I think I'd like to get a better idea of what kind of information we want to expose to the user.  I imagine that the user doesn't want to see rego code and all, but probably wants some sort of actionable info (stats perhaps?).

**Comment by maxsmythe** [comment]
In the constraint framework, the constraint is included as part of the violation.

Client interface:

https://github.com/open-policy-agent/frameworks/blob/347984d8f1064669f54fcbffc04580803082e2dc/constraint/pkg/client/client.go#L24

Individual violation:

https://github.com/open-policy-agent/frameworks/blob/347984d8f1064669f54fcbffc04580803082e2dc/constraint/pkg/types/validation.go#L11

**Comment by maxsmythe** [comment]
Also note that the Rego source code is not part of the constraint, but the template.

For reference:

Example constraint:

https://github.com/open-policy-agent/gatekeeper/blob/1df66f512460b5aa053ea7ed647b98de3f55033e/demo/agilebank/constraints/owner_must_be_provided.yaml#L1

Example template:

https://github.com/open-policy-agent/gatekeeper/blob/1df66f512460b5aa053ea7ed647b98de3f55033e/demo/agilebank/templates/k8srequiredlabels_template.yaml#L1

**Comment by morgante** [comment]
@briantkennedy @t12g I've updated this to simply include the constraint metadata with violations. Please take another look.

File: pkg/gcv/configs/config.go [hunk 3]
```diff
@@ -94,6 +99,23 @@ func asConstraint(data *UnclassifiedConfig) (*Constraint, error) {
 	}, nil
 }
 
+// AsProto returns the constraint a Kubernetes proto
+func (c *Constraint) AsProto() (*validator.Constraint, error) {
+	ci, err := c.Confg.AsInterface()
+	if err != nil {
+		return nil, errors.Wrap(err, "converting to proto")
```
Comment (briantkennedy) [inline pkg/gcv/configs/config.go]: converting bytes to go interface?

Comment (morgante) [reply pkg/gcv/configs/config.go]: That would be the error returned by `Confg.AsInterface`, this appends that the error is in converting to proto.

Comment (briantkennedy) [reply pkg/gcv/configs/config.go]: fine by me

**Pull request merged**